File: hausman_full_reproducibility.R
# ============================================================================
# Hausman Test Monte Carlo -- FULL REPRODUCIBILITY SCRIPT
#
# Reproduces every number reported in the manuscript: the full 192,000-run
# simulation (192 scenarios x 1,000 replications), and the full statistical
# analysis built on it (Tables 1-5 in the published manuscript). This is the
# single script needed to regenerate the paper's results from scratch. Note:
# internal object/file names below (table2, table4, table7, etc.) reflect
# this script's own numbering and do not correspond 1:1 to the manuscript's
# final table numbers after editorial consolidation; see the printed labels
# for what each output actually contains.
#
# Feasibility: the aggregate rank condition for the pooled within-estimator,
# N x (T-1) >= k, holds by at least two orders of magnitude margin for every
# one of the 192 architecture x complexity combinations, so no scenario is
# excluded a priori. (An earlier draft incorrectly excluded scenarios using
# a per-unit T > k rule; that rule was wrong and has been removed. See
# Section 4.1 of the manuscript for the corrected derivation.)
#
# Methodology matches the manuscript exactly (Section 4.5):
#   - ANOVA uses Type II sums of squares (car::Anova), not base R's aov()
#     default (Type I), since Type I is order-dependent and inappropriate
#     for this design.
#   - Pairwise comparisons use two-sample Welch's t-tests (unequal variance),
#     not Bonferroni-corrected tests -- no multiple-comparison correction is
#     applied or claimed.
#   - The architecture x complexity interaction IS included and reported
#     directly, since every combination is now simulated and no cell is
#     structurally excluded.
#   - A logistic regression on the replication-level retain/reject indicator
#     is fit as a robustness check on the ANOVA-on-proportions approach.
#
# Output: written to OUTPUT_DIR. Two categories of files:
#   - Raw simulation results (r_raw_results.csv) -- supplementary
#   - Individual table CSVs -- source data for the manuscript's Tables 1-5
# ============================================================================

suppressMessages({
  library(plm); library(dplyr); library(tidyr); library(purrr); library(car); library(ggplot2)
})

OUTPUT_DIR <- "./results"  # relative to the script location; change if desired
dir.create(OUTPUT_DIR, showWarnings = FALSE, recursive = TRUE)

set.seed(1978)
N_REPLICATIONS <- 1000
ALPHA_LEVEL <- 0.05

panel_architectures <- list(
  "Wide Panel" = list(N = 400, T = 4),
  "Square"     = list(N = 200, T = 8),
  "Long Panel" = list(N = 100, T = 16)
)
variable_complexities <- list(
  "Simple" = 1, "Standard" = 3, "Complex" = 6, "High-Dimensional" = 10
)
missingness_mechanisms <- c("Random", "Early Exit", "Late Missing", "Cyclical")
dropout_rates <- c(0.10, 0.20, 0.30, 0.40)

sig_stars <- function(p) {
  ifelse(is.na(p), "", ifelse(p < 0.001, "***", ifelse(p < 0.01, "**", ifelse(p < 0.05, "*", "ns"))))
}

# ============================================================================
# PART 1: DATA GENERATION AND SIMULATION
# ============================================================================

generate_panel_data <- function(N, T, k) {
  panel_ids <- expand.grid(id = 1:N, time = 1:T)
  alpha <- rnorm(N, mean = 0, sd = 2)
  panel_ids$alpha <- alpha[panel_ids$id]
  x_vars <- matrix(rnorm(N * T * k), nrow = N * T, ncol = k)
  colnames(x_vars) <- paste0("x", 1:k)
  epsilon <- rnorm(N * T, mean = 0, sd = 1.5)
  df <- data.frame(panel_ids, x_vars)
  df$y <- rowSums(df[, paste0("x", 1:k), drop = FALSE]) + df$alpha + epsilon
  df
}

# Missing-data mechanism functions operate on PLAIN data.frame throughout.
# This is the fix: the original implementation carried a plm::pdata.frame
# through these functions, and direct comparison/arithmetic on its pseries
# index columns produced silent errors. Wrapping as pdata.frame happens once,
# immediately before plm::plm() is called.
introduce_random_missing <- function(df, delta) {
  n_obs <- nrow(df)
  missing_indices <- sample(1:n_obs, size = round(delta * n_obs))
  x_cols <- grep("^x", names(df), value = TRUE)
  df[missing_indices, c("y", x_cols)] <- NA
  df
}
introduce_early_exit <- function(df, delta) {
  N <- length(unique(df$id)); T_total <- length(unique(df$time))
  x_cols <- grep("^x", names(df), value = TRUE)
  units_to_drop_count <- max(1, round(2 * N * delta))
  if (delta == 0 || units_to_drop_count == 0) return(df)
  units_to_drop_ids <- sample(unique(df$id), size = units_to_drop_count)
  for (unit_id in units_to_drop_ids) {
    T_exit <- sample(2:T_total, 1)
    df[df$id == unit_id & df$time >= T_exit, c("y", x_cols)] <- NA
  }
  df
}
introduce_late_missing <- function(df, delta) {
  T_total <- length(unique(df$time)); x_cols <- grep("^x", names(df), value = TRUE)
  c_val <- 2 * delta / (T_total + 1)
  df <- df %>% group_by(id) %>%
    mutate(prob_missing = pmin(c_val * time, 0.95), is_missing = runif(n()) < prob_missing) %>%
    ungroup()
  df[df$is_missing, c("y", x_cols)] <- NA
  df$prob_missing <- NULL; df$is_missing <- NULL
  df
}
introduce_cyclical_missing <- function(df, delta) {
  T_total <- length(unique(df$time)); x_cols <- grep("^x", names(df), value = TRUE)
  high_periods <- floor(T_total/2):ceiling(T_total/2 + 1)
  T_high <- length(high_periods); T_low <- T_total - T_high
  if ((3*T_high + T_low) == 0) return(df)
  p_low <- (T_total * delta) / (3*T_high + T_low)
  p_high <- min(3 * p_low, 0.95)
  df$prob <- ifelse(df$time %in% high_periods, p_high, p_low)
  df$is_missing <- runif(nrow(df)) < df$prob
  df[df$is_missing, c("y", x_cols)] <- NA
  df$prob <- NULL; df$is_missing <- NULL
  df
}

run_single_simulation <- function(params) {
  tryCatch({
    complete_data <- generate_panel_data(params$N, params$T, params$k)
    missing_data <- if (params$delta > 0) {
      switch(params$mechanism,
             "Random" = introduce_random_missing(complete_data, params$delta),
             "Early Exit" = introduce_early_exit(complete_data, params$delta),
             "Late Missing" = introduce_late_missing(complete_data, params$delta),
             "Cyclical" = introduce_cyclical_missing(complete_data, params$delta))
    } else complete_data
    obs_per_id <- missing_data %>% filter(!is.na(y)) %>% group_by(id) %>%
      summarise(n_obs = n(), .groups = 'drop')
    valid_ids <- obs_per_id %>% filter(n_obs >= 2) %>% pull(id)  # CORRECTED: a unit needs >=2 observations to contribute any within-unit variation at all; the old n_obs > k threshold was the same per-unit T>k myth applied a second time, post-missingness
    if (length(valid_ids) < 2) {
      return(list(p_value = NA, failure_reason = "Data Failure: Insufficient individuals"))
    }
    final_data <- missing_data %>% filter(id %in% valid_ids) %>%
      plm::pdata.frame(index = c("id", "time"))
    formula <- as.formula(paste("y ~", paste(paste0("x", 1:params$k), collapse = " + ")))
    fe_model <- tryCatch(plm::plm(formula, data = final_data, model = "within"), error = function(e) NULL)
    if (is.null(fe_model)) return(list(p_value = NA, failure_reason = "Model Failure: FE model failed"))
    re_model <- tryCatch(plm::plm(formula, data = final_data, model = "random"), error = function(e) NULL)
    if (is.null(re_model)) return(list(p_value = NA, failure_reason = "Model Failure: RE model failed"))
    ht <- tryCatch(plm::phtest(fe_model, re_model), error = function(e) NULL)
    if (is.null(ht)) return(list(p_value = NA, failure_reason = "Hausman Test Failure"))
    list(p_value = ht$p.value, failure_reason = "Success")
  }, error = function(e) list(p_value = NA, failure_reason = "System Error"))
}

all_scenarios <- expand.grid(
  panel_name = names(panel_architectures), complexity_name = names(variable_complexities),
  mechanism = missingness_mechanisms, delta = dropout_rates, stringsAsFactors = FALSE
) %>% mutate(
  N = map_int(panel_name, ~panel_architectures[[.]]$N),
  T = map_int(panel_name, ~panel_architectures[[.]]$T),
  k = map_int(complexity_name, ~variable_complexities[[.]])
)
feasible_scenarios <- all_scenarios %>% filter(N * (T - 1) >= k)  # CORRECTED: aggregate DOF condition for the pooled within-estimator, not the per-unit T>k heuristic

all_results <- vector("list", nrow(feasible_scenarios))
for (i in 1:nrow(feasible_scenarios)) {
  params <- as.list(feasible_scenarios[i, ])
  scenario_results <- map(1:N_REPLICATIONS, ~ run_single_simulation(params))
  all_results[[i]] <- bind_rows(scenario_results) %>%
    mutate(panel_name = params$panel_name, complexity_name = params$complexity_name,
           mechanism = params$mechanism, delta = params$delta,
           N = params$N, T = params$T, k = params$k,
           specificity = as.numeric(!is.na(p_value) & p_value > ALPHA_LEVEL))
  if (i %% 24 == 0) {
    saveRDS(bind_rows(all_results[1:i]), file.path(OUTPUT_DIR, "r_checkpoint.rds"))
    cat(sprintf("[checkpoint] %d/%d scenarios done\n", i, nrow(feasible_scenarios)))
  }
}
raw <- bind_rows(all_results)
write.csv(raw, file.path(OUTPUT_DIR, "r_raw_results.csv"), row.names = FALSE)
if (file.exists(file.path(OUTPUT_DIR, "r_checkpoint.rds"))) invisible(file.remove(file.path(OUTPUT_DIR, "r_checkpoint.rds")))
cat("Simulation complete:", nrow(raw), "runs.\n\n")

success <- raw %>% filter(failure_reason == "Success")
success$panel_name <- factor(success$panel_name, levels = c("Wide Panel", "Square", "Long Panel"))
success$complexity_name <- factor(success$complexity_name, levels = c("Simple", "Standard", "Complex", "High-Dimensional"))
success$mechanism <- factor(success$mechanism, levels = c("Random", "Early Exit", "Late Missing", "Cyclical"))

# ============================================================================
# PART 2: TABLE 1 -- specificity by mechanism x dropout rate, with per-rate ANOVA
# ============================================================================
table1_rows <- list()
for (d in dropout_rates) {
  row <- list(delta = paste0(d*100, "%"))
  groups <- list()
  for (m in levels(success$mechanism)) {
    sub <- success %>% filter(mechanism == m, delta == d)
    row[[m]] <- sprintf("%.1f%% (n=%d)", mean(sub$specificity)*100, nrow(sub))
    groups[[m]] <- sub$specificity
  }
  fit <- aov(specificity ~ mechanism, data = success %>% filter(delta == d))
  fstat <- summary(fit)[[1]]$`F value`[1]
  pval <- summary(fit)[[1]]$`Pr(>F)`[1]
  row$F_stat <- round(fstat, 2); row$p_value <- round(pval, 3); row$sig <- sig_stars(pval)
  table1_rows[[length(table1_rows)+1]] <- as.data.frame(row, stringsAsFactors = FALSE)
}
table1 <- bind_rows(table1_rows)
write.csv(table1, file.path(OUTPUT_DIR, "table1_mechanism_by_dropout.csv"), row.names = FALSE)

# ============================================================================
# PART 3: TABLE 2 -- overall failure rates by mechanism
# ============================================================================
table2 <- raw %>% group_by(mechanism) %>%
  summarise(total_runs = n(), successful_runs = sum(failure_reason == "Success"),
            failed_runs = total_runs - successful_runs,
            failure_rate = round(failed_runs / total_runs * 100, 2),
            success_rate = round(successful_runs / total_runs * 100, 2), .groups = "drop")
write.csv(table2, file.path(OUTPUT_DIR, "table2_failure_rates.csv"), row.names = FALSE)

# ============================================================================
# PART 4: TABLE 3 -- Long Panel + High-Dimensional outcome breakdown
# ============================================================================
table3 <- raw %>% filter(panel_name == "Long Panel", complexity_name == "High-Dimensional") %>%
  group_by(mechanism, delta) %>%
  summarise(successes = sum(failure_reason == "Success"), n = n(), .groups = "drop")
write.csv(table3, file.path(OUTPUT_DIR, "table3_long_highdim.csv"), row.names = FALSE)

# ============================================================================
# PART 5: TABLE 4 -- specificity by architecture x complexity
# ============================================================================
table4 <- raw %>% group_by(panel_name, complexity_name) %>%
  summarise(N = n(), successful = sum(failure_reason == "Success"),
            failed = N - successful,
            specificity = round(mean(specificity[failure_reason == "Success"]) * 100, 1), .groups = "drop")
write.csv(table4, file.path(OUTPUT_DIR, "table4_architecture_complexity.csv"), row.names = FALSE)

# ============================================================================
# PART 6: TABLE 5 -- 2-way ANOVA (architecture + complexity), Type II SS
# ============================================================================
model5 <- aov(specificity ~ panel_name * complexity_name, data = success)  # UPDATED: includes architecture x complexity interaction, now formally tested per reviewer request
anova5 <- car::Anova(model5, type = 2)
table5 <- data.frame(Effect = rownames(anova5), df = anova5$Df, F_statistic = round(anova5$`F value`, 2),
                      p_value = anova5$`Pr(>F)`, sig = sig_stars(anova5$`Pr(>F)`))
write.csv(table5, file.path(OUTPUT_DIR, "table5_anova_architecture_complexity.csv"), row.names = FALSE)

# ============================================================================
# PART 7: TABLE 6 -- main effects of architecture, with 95% CI
# ============================================================================
table6 <- do.call(rbind, lapply(levels(success$panel_name), function(p) {
  d <- success$specificity[success$panel_name == p]
  m <- mean(d); s <- sd(d); n <- length(d); se <- s / sqrt(n)
  data.frame(panel_name = p, mean_specificity = round(m, 3), sd = round(s, 3), n = n,
             se = round(se, 4), ci_lower = round(m - 1.96*se, 3), ci_upper = round(m + 1.96*se, 3))
}))
write.csv(table6, file.path(OUTPUT_DIR, "table6_main_effects_architecture.csv"), row.names = FALSE)

# ============================================================================
# PART 8: TABLE 7 -- 3-way ANOVA (mechanism + architecture + complexity), Type II SS
# ============================================================================
model7 <- aov(specificity ~ mechanism + factor(delta) + panel_name + complexity_name +
                mechanism:factor(delta) + mechanism:panel_name + mechanism:complexity_name,
              data = success)  # UPDATED: includes mechanism's interactions with rate, architecture, and complexity, per reviewer request
anova7 <- car::Anova(model7, type = 2)
table7 <- data.frame(Effect = rownames(anova7), df = anova7$Df, F_statistic = round(anova7$`F value`, 2),
                      p_value = anova7$`Pr(>F)`, sig = sig_stars(anova7$`Pr(>F)`))
write.csv(table7, file.path(OUTPUT_DIR, "table7_anova_full.csv"), row.names = FALSE)

# ============================================================================
# PART 8b: Logistic regression robustness check on the binary retain/reject outcome
# ============================================================================
model_logit <- glm(specificity ~ mechanism + factor(delta) + panel_name + complexity_name,
                    data = success, family = binomial(link = "logit"))
table7b <- as.data.frame(summary(model_logit)$coefficients)
table7b$term <- rownames(table7b)
write.csv(table7b, file.path(OUTPUT_DIR, "table7b_logistic_robustness.csv"), row.names = FALSE)

# ============================================================================
# PART 9: TABLE 8 -- pairwise comparisons, Welch's t-tests (no correction)
# ============================================================================
pairwise_welch <- function(data, group_col, groups) {
  pairs <- combn(groups, 2, simplify = FALSE)
  do.call(rbind, lapply(pairs, function(pr) {
    da <- data$specificity[data[[group_col]] == pr[1]]
    db <- data$specificity[data[[group_col]] == pr[2]]
    tt <- t.test(da, db)  # Welch's by default (var.equal = FALSE)
    data.frame(comparison = paste(pr[1], "vs", pr[2]),
               p_value = tt$p.value, sig = sig_stars(tt$p.value))
  }))
}
table8_mechanism <- pairwise_welch(success, "mechanism", levels(success$mechanism))
table8_mechanism$factor <- "Missing Data Mechanism"
table8_architecture <- pairwise_welch(success, "panel_name", levels(success$panel_name))
table8_architecture$factor <- "Panel Architecture"
table8_complexity <- pairwise_welch(success, "complexity_name", levels(success$complexity_name))
table8_complexity$factor <- "Model Complexity"
table8 <- bind_rows(table8_mechanism, table8_architecture, table8_complexity) %>%
  select(factor, comparison, p_value, sig)
write.csv(table8, file.path(OUTPUT_DIR, "table8_pairwise_comparisons.csv"), row.names = FALSE)


# ============================================================================
# FIGURES
# ============================================================================
fig_data <- table4
fig_data$panel_name_f <- factor(fig_data$panel_name,
                             levels = c("Long Panel", "Square", "Wide Panel"),
                             labels = c("Long Panel\n(N=100, T=16)", "Square\n(N=200, T=8)", "Wide Panel\n(N=400, T=4)"))
fig_data$complexity_name_f <- factor(fig_data$complexity_name,
                                  levels = c("Simple", "Standard", "Complex", "High-Dimensional"))

# Figure 1: Specificity heatmap (Panel Architecture x Model Complexity)
fig1 <- ggplot(fig_data, aes(x = complexity_name_f, y = panel_name_f, fill = specificity)) +
  geom_tile(color = "white", linewidth = 0.5) +
  geom_text(aes(label = sprintf("%.1f%%", specificity),
                color = specificity < 78),
            fontface = "bold", size = 5.2, show.legend = FALSE) +
  scale_color_manual(values = c(`TRUE` = "white", `FALSE` = "black")) +
  scale_fill_gradientn(colors = c("#c0392b", "#f5f2a8", "#1a6e2e"),
                        limits = c(70, 100), name = "Mean\nSpecificity (%)") +
  labs(x = "Model Complexity", y = "Panel Architecture") +
  theme_minimal(base_size = 13) +
  theme(panel.grid = element_blank(),
        axis.text = element_text(color = "black"),
        axis.title = element_text(size = 13))
ggsave(file.path(OUTPUT_DIR, "figure1_specificity_heatmap.png"), fig1,
       width = 8.2, height = 4.6, dpi = 300, bg = "white")

# Figure 2: Specificity bar chart (Model Complexity x Panel Architecture)
fig_data$panel_name_short <- factor(fig_data$panel_name,
  levels = c("Wide Panel", "Square", "Long Panel"))
fig_data$complexity_short <- factor(fig_data$complexity_name,
  levels = c("Simple", "Standard", "Complex", "High-Dimensional"),
  labels = c("Simple", "Standard", "Complex", "High-Dim."))

fig2 <- ggplot(fig_data, aes(x = complexity_short, y = specificity, fill = panel_name_short)) +
  geom_col(position = position_dodge(width = 0.8), width = 0.75) +
  geom_text(aes(label = sprintf("%.1f%%", specificity)),
            position = position_dodge(width = 0.8), vjust = -0.4, angle = 90,
            fontface = "bold", size = 3.4, hjust = 0) +
  scale_fill_manual(values = c("Wide Panel" = "#4472C4", "Square" = "#548235", "Long Panel" = "#C00000")) +
  scale_y_continuous(limits = c(0, 108), breaks = seq(0, 100, 20)) +
  labs(x = NULL, y = "Mean Specificity (%)", fill = NULL) +
  theme_minimal(base_size = 13) +
  theme(legend.position = "top",
        panel.grid.minor = element_blank(),
        panel.grid.major.x = element_blank(),
        axis.text = element_text(color = "black"))
ggsave(file.path(OUTPUT_DIR, "figure2_specificity_bars.png"), fig2,
       width = 8.2, height = 4.8, dpi = 300, bg = "white")

cat("Both figures saved to", OUTPUT_DIR, "\n\n")

# ============================================================================
# DONE
# ============================================================================
cat("All tables written to", OUTPUT_DIR, "\n\n")
cat("--- Table 2 (mechanism failure rates) ---\n"); print(table2)
cat("\n--- Table 5 (2-way ANOVA) ---\n"); print(table5)
cat("\n--- Table 7 (3-way ANOVA) ---\n"); print(table7)

File: hausman_full_reproducibility.py
"""
============================================================================
Hausman Test Monte Carlo -- INDEPENDENT PYTHON REIMPLEMENTATION

Independently reproduces the manuscript's core simulation: the full 192,000-
run design (192 scenarios x 1,000 replications). Shares no code with the R
implementation (hausman_full_reproducibility.R); built from the manuscript's
equations directly, using pandas and linearmodels rather than plm.

Feasibility: the aggregate rank condition N x (T-1) >= k holds by at least
two orders of magnitude margin for every one of the 192 architecture x
complexity combinations, so no scenario is excluded a priori.

Output: raw_results_python.csv, one row per replication, matching the
column structure of the R implementation's r_raw_results.csv, so results
from both languages can be pooled or compared directly.
============================================================================
"""
import numpy as np
import pandas as pd
from linearmodels.panel import PanelOLS, RandomEffects
from scipy import stats
import time

OUTPUT_DIR = "./results"
import os
os.makedirs(OUTPUT_DIR, exist_ok=True)

SEED = 1978
N_REPLICATIONS = 1000  # matches the R implementation; reduce for a quick test run

PANEL_ARCHITECTURES = {
    "Wide Panel": (400, 4),
    "Square": (200, 8),
    "Long Panel": (100, 16),
}
COMPLEXITY_LEVELS = {
    "Simple": 1, "Standard": 3, "Complex": 6, "High-Dimensional": 10,
}
MECHANISMS = ["Random", "Early Exit", "Late Missing", "Cyclical"]
DROPOUT_RATES = [0.10, 0.20, 0.30, 0.40]


def generate_panel(N, T, k, rng):
    ids = np.repeat(np.arange(N), T)
    times = np.tile(np.arange(1, T + 1), N)  # 1-indexed, matches manuscript's t = 1,...,T
    alpha = rng.normal(0, 2, N)              # sd=2 -> Var=4
    alpha_long = np.repeat(alpha, T)
    X = rng.normal(0, 1, size=(N * T, k))
    eps = rng.normal(0, 1.5, N * T)          # sd=1.5 -> Var=2.25
    beta = np.ones(k)
    Y = X @ beta + alpha_long + eps
    df = pd.DataFrame(X, columns=[f"x{i+1}" for i in range(k)])
    df["id"] = ids
    df["time"] = times
    df["y"] = Y
    return df


def apply_missingness(df, mechanism, delta, x_cols, rng):
    d = df.copy()
    T_total = d["time"].nunique()
    N = d["id"].nunique()

    if mechanism == "Random":
        mask = rng.random(len(d)) < delta
        d.loc[mask, ["y"] + x_cols] = np.nan

    elif mechanism == "Early Exit":
        n_drop = max(1, round(2 * N * delta))
        if delta > 0 and n_drop > 0:
            unit_ids = d["id"].unique()
            drop_ids = rng.choice(unit_ids, size=n_drop, replace=False)
            for uid in drop_ids:
                t_exit = rng.integers(2, T_total + 1)
                mask = (d["id"] == uid) & (d["time"] >= t_exit)
                d.loc[mask, ["y"] + x_cols] = np.nan

    elif mechanism == "Late Missing":
        c_val = 2 * delta / (T_total + 1)
        prob = np.minimum(c_val * d["time"], 0.95)
        mask = rng.random(len(d)) < prob
        d.loc[mask, ["y"] + x_cols] = np.nan

    elif mechanism == "Cyclical":
        lo, hi = int(np.floor(T_total / 2)), int(np.ceil(T_total / 2 + 1))
        high_periods = set(range(lo, hi + 1))
        T_high = len(high_periods)
        T_low = T_total - T_high
        p_low = T_total * delta / (3 * T_high + T_low)
        p_high = min(3 * p_low, 0.95)
        probs = d["time"].apply(lambda t: p_high if t in high_periods else p_low)
        mask = rng.random(len(d)) < probs
        d.loc[mask, ["y"] + x_cols] = np.nan

    return d


def hausman_test(df, x_cols):
    d = df.dropna(subset=["y"] + x_cols).set_index(["id", "time"])
    if d.empty or d.reset_index()["id"].nunique() < 2:
        return None, "insufficient_data"
    try:
        exog = d[x_cols]
        fe_res = PanelOLS(d["y"], exog, entity_effects=True).fit()
        re_res = RandomEffects(d["y"], exog).fit()
    except Exception:
        return None, "estimation_error"
    diff = re_res.params - fe_res.params
    var_diff = fe_res.cov - re_res.cov
    try:
        inv_var_diff = np.linalg.inv(var_diff.values)
    except np.linalg.LinAlgError:
        return None, "singular_matrix"
    H = diff.values @ inv_var_diff @ diff.values
    if H < 0:
        # Finite-sample noise can push the estimated variance difference
        # slightly off positive-semi-definite; clip to the boundary (H=0)
        # rather than treat as a computational failure, consistent with
        # standard practice for this known edge case.
        H = 0.0
    p_value = 1 - stats.chi2.cdf(H, len(x_cols))
    return p_value, "success"


def run_simulation():
    # --- Checkpoint / resume setup ---
    # Output path is resolved to an absolute path up front, and a checkpoint
    # file tracks which scenarios are already done. If this script is
    # interrupted (power loss, USB/network drive disconnect, crash, etc.),
    # rerunning it picks up where it left off instead of starting over.
    output_dir_abs = os.path.abspath(OUTPUT_DIR)
    os.makedirs(output_dir_abs, exist_ok=True)
    results_path = os.path.join(output_dir_abs, "raw_results_python.csv")
    checkpoint_path = os.path.join(output_dir_abs, "_checkpoint.txt")

    print(f"Writing results to: {results_path}")
    if os.path.abspath(os.getcwd()).split(os.sep)[0] != output_dir_abs.split(os.sep)[0]:
        print("NOTE: output drive differs from current drive; this is fine, "
              "just confirming the absolute path above is where your data will land.")

    completed_scenarios = set()
    if os.path.exists(checkpoint_path):
        with open(checkpoint_path, "r") as f:
            completed_scenarios = set(line.strip() for line in f if line.strip())
        print(f"Resuming: {len(completed_scenarios)} scenario(s) already completed "
              f"per checkpoint file.")

    file_exists = os.path.exists(results_path) and len(completed_scenarios) > 0

    rng = np.random.default_rng(SEED)
    total_scenarios = len(PANEL_ARCHITECTURES) * len(COMPLEXITY_LEVELS) * len(MECHANISMS) * len(DROPOUT_RATES)
    scenario_num = 0
    start = time.time()

    for panel_name, (N, T) in PANEL_ARCHITECTURES.items():
        for complexity_name, k in COMPLEXITY_LEVELS.items():
            x_cols = [f"x{i+1}" for i in range(k)]
            for mechanism in MECHANISMS:
                for delta in DROPOUT_RATES:
                    scenario_num += 1
                    scenario_key = f"{panel_name}|{complexity_name}|{mechanism}|{delta}"

                    if scenario_key in completed_scenarios:
                        continue  # already done in a prior run; skip but keep RNG state moving on is not needed since skipped scenarios don't consume draws here

                    rows = []
                    for rep in range(N_REPLICATIONS):
                        df = generate_panel(N, T, k, rng)
                        d = apply_missingness(df, mechanism, delta, x_cols, rng)
                        p_value, failure_reason = hausman_test(d, x_cols)
                        specificity = int(p_value > 0.05) if p_value is not None else np.nan
                        rows.append({
                            "p_value": p_value,
                            "failure_reason": failure_reason,
                            "panel_name": panel_name,
                            "complexity_name": complexity_name,
                            "mechanism": mechanism,
                            "delta": delta,
                            "N": N, "T": T, "k": k,
                            "specificity": specificity,
                        })

                    # Append this scenario's results to disk immediately.
                    scenario_df = pd.DataFrame(rows)
                    scenario_df.to_csv(results_path, mode="a", header=not file_exists, index=False)
                    file_exists = True

                    # Record this scenario as done, immediately, on disk.
                    with open(checkpoint_path, "a") as f:
                        f.write(scenario_key + "\n")

                    elapsed = time.time() - start
                    print(f"Scenario {scenario_num}/{total_scenarios} done "
                          f"({elapsed/60:.1f} min elapsed) -- saved to disk")

    print(f"\nDone. All {total_scenarios} scenarios written to {results_path}")
    print("You can delete the _checkpoint.txt file now that the run is complete.")
    return pd.read_csv(results_path)


if __name__ == "__main__":
    cwd = os.path.abspath(os.getcwd())
    if len(cwd) >= 2 and cwd[1] == ':' and cwd[0].upper() not in ('C',):
        print(f"WARNING: running from drive {cwd[0]}:, which may be removable media")
        print("(a USB or external drive). If this drive is disconnected while")
        print("the script is running, the run WILL be interrupted, though progress")
        print("is saved after every scenario and can be resumed by rerunning.")
        print("Recommended: copy this script to your local hard drive (e.g. C:)")
        print("and run it from there instead.\n")
    run_simulation()
